src/nec-usbatermfw.c

```diff
@@ -0,0 +1,328 @@
+	}
+
+	return padlen;
+}
+
+static int write_blockheader(uint32_t flags,
+			     uint32_t loadaddr, uint32_t entryp,
+			     size_t datalen, int padlen, uint32_t cksum)
+{
+	struct header hdr;
+	void *cur = &hdr;
+	char buf[0x20];
+
+	flags <<= 16;
+	flags |= (~flags & 0xffff0000) >> 16;
+
+	hdr.flags = HOST_TO_BE32(flags);
+	hdr.totlen = HOST_TO_BE32(BLKHDR_LEN + datalen);
+	hdr.hdrlen = HOST_TO_BE32(BLKHDR_LEN);
+	hdr.cksum = 0;
+	hdr.loadaddr = HOST_TO_BE32(loadaddr);
+	hdr.entryp = HOST_TO_BE32(entryp);
+
+	for (cur += 4; cur - (void *)&hdr < BLKHDR_LEN; cur += 2) {
+		cksum += HOST_TO_LE16(*(uint16_t *)cur);
+		/*
+		 * workaround of unknown bug if built with gcc 9.4.0
+		 * (Ubuntu 9.4.0-1ubuntu1~20.04.2) and cmake
+		 */
+		snprintf(buf, sizeof(buf), "%u", cksum);
+	}
+
+	cksum = 0xffff ^ cksum % 0xffff;
+	hdr.cksum = HOST_TO_LE32(cksum);
+
+	if (datalen)
+		fseek(outbin, -(BLKHDR_LEN + datalen + padlen), SEEK_CUR);
+	if (fwrite(&hdr, 1, BLKHDR_LEN, outbin) != BLKHDR_LEN) {
+		fprintf(stderr, "Couldn't write %d bytes\n", BLKHDR_LEN);
+		return -EACCES;
+	}
+	if (datalen)
+		fseek(outbin, datalen + padlen, SEEK_CUR);
+
+	nblk++;
+	printf("%d:\t0x%04x,\t0x%08zx,\t0x%08x,\t0x%08x,\t0x%08x\n",
+	       nblk, flags >> 16, BLKHDR_LEN + datalen, loadaddr, entryp, cksum);
+	return 0;
+}
+
+static int append_datablock_from_file(uint32_t flags,
+			 uint32_t loadaddr, uint32_t entryp,
+			 const char *datapath)
+{
+	FILE *databin;
+	size_t readlen, length = 0;
+	uint32_t cksum = 0;
+	char buf[0x10000];
+	int ret = 0;
+
+	fseek(outbin, BLKHDR_LEN, SEEK_CUR);
+
+	databin = fopen(datapath, "r");
+	if (!databin) {
+		fprintf(stderr, "couldn't open %s\n", datapath);
+		return -EACCES;
+	}
+
+	while ((readlen = fread(buf, 1, sizeof(buf), databin)) > 0) {
+		ret = write_data(buf, readlen, &cksum);
+		if (ret < 0)
+			goto exit;
+		length += readlen;
+	}
+
+	ret = write_blockheader(flags, loadaddr, entryp, length, ret, cksum);
+
+exit:
+	fclose(databin);
+	return ret;
+}
+
+static int append_datablock_from_buf(uint32_t flags,
+				     uint32_t loadaddr, uint32_t entryp,
+				     const char *buf, size_t buflen)
+{
+	uint32_t cksum = 0;
+	int ret;
+
+	fseek(outbin, BLKHDR_LEN, SEEK_CUR);
+
+	ret = write_data(buf, buflen, &cksum);
+
+	return write_blockheader(flags, loadaddr, entryp, buflen, ret, cksum);
+}
+
+int main(int argc, char **argv)
+{
+	uint32_t flags = BLKHDR_FLAGS_EXEC;
+	uint32_t loadaddr = BLKHDR_DEF_LOADADDR;
+	uint32_t entryp = BLKHDR_DEF_ENTRYP;
+	char buf[0x40], ftype[FILETYPE_LEN];
+	int ret, c, ftlen = 0;
+
+	progname = basename(argv[0]);
+
+	if (argc >= 2 &&
+	    (!strcmp(argv[1], "-h") || !strcmp(argv[1], "--help"))) {
+		usage();
+		return 0;
+	}
+
+	if (argc < 2) {
+		fprintf(stderr, "no output file specified\n");
+		usage();
+		return -EINVAL;
+	}
+
+	outbin = fopen(argv[1], "w+");
+	if (!outbin) {
+		fprintf(stderr, "Couldn't open %s\n", argv[1]);
+		return -EACCES;
+	}
+
+	while((c = getopt(argc, argv, "f:a:e:d:t:")) != -1) {
```

Comment:
These two `getopt` loops are strange. 

I would prefer if you put this into one `getopt` loop.

Reply:
OK, I'll update...